yarn-project/merkle-tree/src/sparse_tree/sparse_tree.ts

```diff
@@ -0,0 +1,298 @@
+  }
+
+  /**
+   * Sets the root, depth and size of the tree based on the meta data which are associated with the tree name.
+   */
+  public async syncFromDb() {
+    const meta: Buffer | undefined = await this.dbGet(this.name);
+    if (!meta) {
+      return;
+    }
+    const { root, depth, size } = decodeMeta(meta);
+    this.root = root;
+    this.depth = depth;
+    this.size = size;
+    this.clearCache();
+  }
+
+  /**
+   * Returns the root of the tree.
+   * @returns The root of the tree.
+   */
+  public getRoot(includeUncommitted = false): Buffer {
+    return !includeUncommitted ? this.root : this.cache[indexToKeyHash(this.name, 0, 0n)] ?? this.root;
+  }
+
+  /**
+   * Returns the number of leaves in the tree.
+   * @returns The number of leaves in the tree.
+   */
+  public getNumLeaves(includeUncommitted = false) {
+    return !includeUncommitted ? this.size : this.cachedSize ?? this.size;
+  }
+
+  /**
+   * Returns the name of the tree.
+   * @returns The name of the tree.
+   */
+  public getName(): string {
+    return this.name;
+  }
+
+  /**
+   * Returns the depth of the tree.
+   * @returns The depth of the tree.
+   */
+  public getDepth(): number {
+    return this.depth;
+  }
+
+  /**
+   * Returns a sibling path for the element at the given index.
+   * @param index - The index of the element.
+   * @returns A sibling path for the element at the given index.
+   * Note: The sibling path is an array of sibling hashes, with the lowest hash (leaf hash) first, and the highest hash last.
+   */
+  public async getSiblingPath(index: bigint, includeUncommitted = false) {
+    const path = new SiblingPath();
+    let level = this.depth;
+    while (level > 0) {
+      const isRight = index & 0x01n;
+      const sibling = await this.getLatestValueAtIndex(level, isRight ? index - 1n : index + 1n, includeUncommitted);
+      path.data.push(sibling);
+      level -= 1;
+      index >>= 1n;
+    }
+    return path;
+  }
+
+  /**
+   * Appends the given leaves to the tree.
+   * @param leaves - The leaves to append.
+   * @returns Empty promise.
+   */
+  public appendLeaves(leaves: Buffer[]): Promise<void> {
+    throw Error('Not implemented');
+  }
+
+  /**
+   * Updates a leaf in the tree.
+   * @param leaf - New contents of the leaf.
+   * @param index - Index of the leaf to be updated.
+   */
+  public async updateLeaf(leaf: Buffer, index: bigint): Promise<void> {
+    if (index > this.maxIndex) {
+      throw Error(`Index out of bounds. Index ${index}, max index: ${this.maxIndex}.`);
+    }
+    const insertingEmpty = leaf.equals(SparseMerkleTree.ZERO_ELEMENT);
+    const originallyEmpty = (await this.getLeafValue(index, true))?.equals(SparseMerkleTree.ZERO_ELEMENT);
+    if (insertingEmpty && originallyEmpty) {
+      return;
+    }
+    await this.addLeafToCacheAndHashToRoot(leaf, index);
+    if (insertingEmpty) {
+      this.cachedSize!--;
```

Comment:
I don't think there is any guarantee that cachedSize !== undefined here.

Reply:
You are right. Good catch. Addressed in 7aecd5e2733c5c828452e269e4adc9b13c376954